avaframe/com1DFAPy/SPHfunctions.py

```diff
@@ -0,0 +1,408 @@
+        iPend = indPartInCell[min(ic+3, ncols * (indy + n + 1))]
+        # loop on all particles in neighbour boxes
+        for p in range(iPstart, iPend):
+            # index of particle in neighbour box
+            l = int(partInCell[p])
+            if j != l:
+                L = np.append(L, l)
+                dx = particles['x'][l] - x
+                dy = particles['y'][l] - y
+                dz = particles['z'][l] - z
+                r = DFAtls.norm(dx, dy, dz)
+                if r < 0.001 * rKernel:
+                    # impose a minimum distance between particles
+                    r = 0.001 * rKernel
+                if r < rKernel:
+                    hr = rKernel - r
+                    dwdr = dfacKernel * hr * hr
+                    massl = particles['m'][l]
+                    gradhX = gradhX + massl * dwdr * dx / r
+                    gradhY = gradhY + massl * dwdr * dy / r
+                    gradhZ = gradhZ + massl * dwdr * dz / r
+    # print((time.time() - startTime)/1)
+
+    return gradhX, gradhY,  gradhZ, L
+
+
+def calcGradHSPHVect(particles, j, ncols, nrows, csz):
+    # SPH kernel
+    # use "spiky" kernel: w = (h - r)**3 * 10/(pi*h**5)
+    rKernel = csz
+    facKernel = 10.0 / (math.pi * pow(rKernel, 5.0))
+    dfacKernel = -3.0 * facKernel
+
+    indx, indy, _ = particles['InCell'][j]
+    indPartInCell = particles['indPartInCell']
+    partInCell = particles['partInCell']
+    x = particles['x'][j]
+    y = particles['y'][j]
+    z = particles['z'][j]
+    gradhX = 0
+    gradhY = 0
+    gradhZ = 0
+    # no loop option
+
+    # startTime = time.time()
+    # find all the neighbour boxes
+    # index of the left box
+    # check if we are on the bottom ot top row!!!
+    lInd = -1
+    rInd = 2
+    if indy == 0:
+        lInd = 0
+    if indy == nrows - 1:
+        rInd = 1
+    ic = (indx - 1) + ncols * (indy + np.arange(lInd, rInd, 1))
+    # go throught all index from left middle and right box and get the
+    # particles in those boxes
+    # make sure not to take particles from the other edge
+    iPstart = indPartInCell[np.maximum(ic, ncols * (indy + np.arange(lInd, rInd, 1)))]
+    iPend = indPartInCell[np.minimum(ic + 3, ncols * (indy + np.arange(lInd, rInd, 1) + 1))]
+
+    # gather all the particles
+    index = np.concatenate([np.arange(x, y) for x, y in zip(iPstart, iPend)])
+
+    # compute SPH gradient
+    L = partInCell[index]
+    # make sure to remove the j particle
+    indJ = np.where(L == j)
+    L = np.delete(L, indJ)
+    dx = particles['x'][L] - x
+    dy = particles['y'][L] - y
+    dz = (particles['z'][L] - z)
+    r = DFAtls.norm(dx, dy, dz)
+    # impose a minimum distance between particles
+    dx = np.where(r < 0.0001 * rKernel, 0.0001 * rKernel * dx, dx)
+    dy = np.where(r < 0.0001 * rKernel, 0.0001 * rKernel * dy, dy)
+    dz = np.where(r < 0.0001 * rKernel, 0.0001 * rKernel * dz, dz)
+    r = np.where(r < 0.0001 * rKernel, 0.0001 * rKernel, r)
```

Comment:
Should we put this distance somewhere as parameter, or is 0.0001 a commonly accepted value?

Reply:
also created one yesterday

Reply:
# 204